code/modules/reagents/newchem/chem_heater.dm

```diff
@@ -0,0 +1,114 @@
+/obj/machinery/chem_heater
+	name = "chemical heater"
+	density = 1
+	anchored = 1
+	icon = 'icons/obj/chemical.dmi'
+	icon_state = "mixer0b"
+	use_power = 1
+	idle_power_usage = 40
+	var/obj/item/weapon/reagent_containers/beaker = null
+	var/temperature = 300
+	var/rate = 10 //heating/cooling rate, default is 10 kelvin per tick
+	var/on = FALSE
+
+
+/obj/machinery/chem_heater/process()
+	..()
+	if(stat & NOPOWER)
+		return
+	var/state_change = 0
+	if(on)
+		if(beaker)
+			if(beaker.reagents.chem_temp > temperature)
+				beaker.reagents.chem_temp = max(beaker.reagents.chem_temp-rate, temperature)
+				beaker.reagents.handle_reactions()
+				state_change = 1
+			else if(beaker.reagents.chem_temp < temperature)
+				beaker.reagents.chem_temp = min(beaker.reagents.chem_temp+rate, temperature)
+				beaker.reagents.handle_reactions()
+				state_change = 1
+	if(state_change)
+		nanomanager.update_uis(src)
+
+/obj/machinery/chem_heater/proc/eject_beaker()
+	if(beaker)
+		beaker.loc = get_turf(src)
+		beaker.reagents.handle_reactions()
+		beaker = null
+		icon_state = "mixer0b"
+		nanomanager.update_uis(src)
+
+/obj/machinery/chem_heater/power_change()
+	if(powered())
+		stat &= ~NOPOWER
+	else
+		spawn(rand(0, 15))
+			stat |= NOPOWER
+	nanomanager.update_uis(src)
+
+/obj/machinery/chem_heater/attackby(var/obj/item/I as obj, var/mob/user as mob)
+	if(isrobot(user))
+		return
+
+	if(istype(I, /obj/item/weapon/reagent_containers/glass))
+		if(beaker)
+			user << "<span class='notice'>A beaker is already loaded into the machine.</span>"
+			return
+
+		if(user.drop_item())
+			beaker = I
+			I.loc = src
+			user << "<span class='notice'>You add the beaker to the machine!</span>"
+			icon_state = "mixer1b"
+			nanomanager.update_uis(src)
+
+/obj/machinery/chem_heater/attack_hand(var/mob/user as mob)
+	ui_interact(user)
+
+/obj/machinery/chem_heater/Topic(href, href_list)
+	if(..())
+		return 0
+
+	if(href_list["toggle_on"])
+		on = !on
+		. = 1
+
+	if(href_list["adjust_temperature"])
+		var/val = href_list["adjust_temperature"]
+		if(isnum(val))
+			temperature = Clamp(temperature+val, 0, 1000)
+		else if(val == "input")
+			temperature = Clamp(input("Please input the target temperature", name) as num, 0, 1000)
```

Comment:
that reminds me i forgot to make CLF3 shoot the beaker temp up to 1000k, that's why I have the max set as that